File: Queue/Double Ended Queue/Double Ended Insertion Queue/DEIQ.c
#include <stdio.h>

// Queue Initialization
#define MAX 5
int queue[MAX];
int BACK=-1, FRONT=-1;

enum condition{FALSE=-1, TRUE=1};
enum mode{front, back};

// OVERFLOW CONDITION
int OVERFLOW(int ForB) {
	if (ForB == front) {
		if (BACK == MAX-1) {
			return TRUE;
		}
		else {
			return FALSE;
		}
	}
	else {
		if (FRONT <= 0) {
			return TRUE;
		}
		else {
			return FALSE;
		}
	}
}

// UNDERFLOW CONDITION
int UNDERFLOW() {
    if (BACK == -1) {
        return TRUE;
    } 
    else {
        return FALSE;
    }
}

// Push into queue BACK
int insert_front(int item) {
    if (OVERFLOW(front) == FALSE) {
        if (BACK==-1)
            FRONT=0;
        queue[++BACK] = item;
        return TRUE;
    }
    else {
        return FALSE;
    }
}

// Push into queue FRONT
int insert_back(int item) {
    if (OVERFLOW(back) == FALSE) {
        queue[--FRONT] = item;
        return TRUE;
    }
    else {
        return FALSE;
    }
}

// Pop from queue FRONT
int delete() {
    if (UNDERFLOW() == FALSE) {
		int temp = queue[FRONT];
		if (FRONT == BACK)
			FRONT = BACK = -1;
		else
			FRONT++;
        return temp;
    }
    else {
        return -1;
    }
}

// Display item in queue
void display() {
    if (UNDERFLOW() == FALSE) {
        for (int i=FRONT; i<=BACK; i++) {
            printf("%-5d", queue[i]);
        }
        printf("\n");
    }
    else {
        printf("-> UNDERFLOW <-\n");
    }
}


// DRIVER CODE
int main() {
    int switch_var;		// Switch Variable
	int exit = FALSE;	// Exit condition
	int temp;			// Temporary variable

	while (exit == FALSE) {
		// Menu
		printf("\nEnter \n\t1. INSERT BACK\n\t2. INSERT FRONT \n\t3. DELETE \n\t4. DISPLAY \n\t5. EXIT \n-> ");
		scanf("%d", &switch_var);

		switch (switch_var) {
			// INSERT BACK
			case 1:
				printf("Enter a value to insert : ");
				scanf("%d", &temp);
				if (insert_front(temp) == TRUE) {  // If not oveflowed
					printf("%d is inserted from front into queue.\n", temp);
				}
				else {
					printf("-> OVERFLOW <- \n");
				}
				break;
			// INSERT FRONT
			case 2:
				printf("Enter a value to insert : ");
				scanf("%d", &temp);
				if (insert_back(temp) == TRUE) {  // If not oveflowed
					printf("%d is inserted from back into queue.\n", temp);
				}
				else {
					printf("-> OVERFLOW <- \n");
				}
				break;
			// DELETE
			case 3:
				temp = delete();
				if (temp == FALSE) {  // If not underflowed
					printf("-> UNDERFLOW <- \n");
				}
				else {
					printf("The deleted value is %d.\n", temp);
				}
				break;
			// DISPLAY
			case 4:
				display();
				break;
			// EXIT
			case 5:
				exit = TRUE;
				break;
			// INVALID
			default:
				printf("Enter a valid option\n");
				break;
		}
	}
	
	return 0;
}

File: Data-Structure/Queue/Double Ended Queue/Double Ended Insertion Queue/DEIQ.py
class Queue():
    # Queue Initialization
    def __init__(self):
        self.MAX = 5
        self.queue = []

    # OVERFLOW CONDITION
    def OVERFLOW(self):
        if len(self.queue) == self.MAX:
            return True
        else:
            return False

    # UNDERFLOW CONDITION
    def UNDERFLOW(self):
        if len(self.queue) == 0:
            return True
        else:
            return False
    
    # Insert into queue FRONT
    def insert_front(self, item):
        if not self.OVERFLOW():
            self.queue.insert(0, item)
            return True
        else:
            return False

    # Insert into queue BACK
    def insert_back(self, item):
        if not self.OVERFLOW():
            self.queue.append(item)
            return True
        else:
            return False
    
    # Delete from queue FRONT
    def delete(self):
        if not self.UNDERFLOW():
            return self.queue.pop(0)
        else:
            return False
    
    def display(self):
        if not self.UNDERFLOW():
            print("   ".join([str(i) for i in self.queue]))
        else:
            print("-> UNDERFLOW <-")

menu = '''
Enter
    1. INSERT FRONT
    2. INSERT BACK
    3. DELETE
    4. DISPAY
    5. EXIT
-> '''

if __name__ == "__main__":
    obj = Queue()
    exit = False
    while not exit:
        switch_var = int(input(menu))
        # INSERT FRONT
        if switch_var == 1:
            temp = int(input("Enter the number : "))
            if obj.insert_front(temp):
                print(f"{temp} is inserted into queue from front.")
            else:
                print("-> OVERFLOW <-")
        # INSERT BACK
        elif switch_var == 2:
            temp = int(input("Enter the number : "))
            if obj.insert_back(temp):
                print(f"{temp} is inserted into queue fron back.")
            else:
                print("-> OVERFLOW <-")
        # DELETE 
        elif switch_var == 3:
            temp = obj.delete()
            if temp:
                print(f"{temp} is deleted from queue.")
            else:
                print("-> UNDERFLOW <-")
        # DISPLAY
        elif switch_var == 4:
            obj.display()
        # EXIT
        elif switch_var == 5:
            exit = True
        # INVALID
        else:
            print("Please enter a valid statement.")
